portal-impl/src/com/liferay/portlet/asset/service/impl/AssetCategoryLocalServiceImpl.java

```diff
@@ -415,6 +430,52 @@ public List<AssetCategory> getEntryCategories(long entryId) {
 		return Collections.emptyList();
 	}
 
+	@Override
+	public AssetCategory getOrAddIncompleteCategory(
+			String externalReferenceCode, long userId, long groupId)
+		throws PortalException {
+
+		AssetCategory assetCategory = fetchAssetCategoryByExternalReferenceCode(
+			externalReferenceCode, groupId);
+
+		if (assetCategory != null) {
+			return assetCategory;
+		}
+
+		if (!LazyReferencingThreadLocal.isEnabled()) {
+			throw new NoSuchCategoryException(
+				StringBundler.concat(
+					"Unable to find asset category with external reference ",
+					"code ", externalReferenceCode, " and group ", groupId));
+		}
+
+		assetCategory = assetCategoryLocalService.fetchCategory(
+			groupId, AssetCategoryConstants.INCOMPLETE_PARENT_CATEGORY_ID,
+			externalReferenceCode,
+			AssetVocabularyConstants.INCOMPLETE_VOCABULARY_ID);
+
+		if (assetCategory != null) {
+			throw new DuplicateCategoryException(
+				StringBundler.concat(
+					"There is another category named ", externalReferenceCode,
```

Comment:
the previous error message is "...asset category...", this one "...category..."

Reply:
as pointed by @marcogalluzzi , this exception should be thrown by the add* method so I will remove this check